### PR TITLE
Merge 2.28.0 back to mainline

### PR DESCRIPTION
**Pull request by sreuland**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://developers.stellar.org/api/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

merging 2.28.0 back to mainline, it has a few material changes that were done solely on release branch to start such as #5171 and #5175. And there was a significant conflict in histoaryarchive package between the branches as master had taken a refactoring from lighthorizin and 2.28. had a somewhat partial/different refactoring in the same package. i updated all to be consistent to the historyarchive package from master. 

### Why

part of post release steps

### CODE REVIEW

File: historyarchive/archive_pool.go [hunk 1]
```diff
@@ -51,8 +51,18 @@ func NewArchivePool(archiveURLs []string, opts ArchiveOptions) (ArchivePool, err
 	return validArchives, nil
 }
 
+func (pa ArchivePool) GetStats() []ArchiveStats {
+	stats := []ArchiveStats{}
```
Comment (Shaptic) [inline historyarchive/archive_pool.go:55]: we can pre-allocate the known capacity here, yeah?
```suggestion
	stats := make([]ArchiveStats, 0, len(pa))
```

File: historyarchive/archive_pool.go [hunk 2]
```diff
@@ -51,8 +51,18 @@ func NewArchivePool(archiveURLs []string, opts ArchiveOptions) (ArchivePool, err
...
+	for _, archive := range pa {
+		if len(archive.GetStats()) == 1 {
```
Comment (Shaptic) [inline historyarchive/archive_pool.go:57]: bit confused about the check here?

File: ingest/ledgerbackend/captive_core_backend_test.go [hunk 3]
```diff
@@ -138,22 +140,33 @@ func TestCaptiveNew(t *testing.T) {
 	require.NoError(t, err)
 	defer os.RemoveAll(storagePath)
 
+	var userAgent string
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		userAgent = r.Header["User-Agent"][0]
```
Comment (Shaptic) [inline ingest/ledgerbackend/captive_core_backend_test.go:145]: just checking: why `[0]` here?

**Review by Shaptic — APPROVED** [review]
looks like a clean merge to me - thanks for resolving all of the conflicts!! :clap:

File: historyarchive/archive_pool.go (hunk 2 above)
Comment (sreuland) [reply historyarchive/archive_pool.go]: i think was left over from initial impl, it works out now to just rollup whatever's in the array, good catch - https://github.com/stellar/go/pull/5183/commits/98d54d1623e2bd3f1370061b40fb04aa53dcc03c

File: historyarchive/archive_pool.go (hunk 1 above)
Comment (sreuland) [reply historyarchive/archive_pool.go:55]: I think potential capacity on stats isn't [one-to-one irt archive](https://github.com/stellar/go/pull/5183/commits/98d54d1623e2bd3f1370061b40fb04aa53dcc03c), there could be more than one stats instance, not at present, it doesn't seem like big gain vs. code either way?

File: ingest/ledgerbackend/captive_core_backend_test.go (hunk 3 above)
Comment (sreuland) [reply ingest/ledgerbackend/captive_core_backend_test.go:145]: the expectation is that only a single value for user-agent header is sent by the archive client when config'd correctly, the test should break if anything different emitted, i.e. multi or no value.

Comment (Shaptic) [reply ingest/ledgerbackend/captive_core_backend_test.go:145]: got it, thanks for clarity!

**Pull request merged**